keybase/keybase-chat-bot
# 🤓 Functional Restructure

### Goals

The purpose of restructuring is to transition the module into a general keybase bot as outlined in #18. The approach taken draws from functional programming with the hopes of improving usability, testing, and development.

- Improve future unit + integration testing by refactoring each API into functions rather than classes
- Make bot state transitions predictable and pure with a "redux"-like store
- Writing tests for multiple concurrent bots should be easy.
- Spinning up a bot into any possible state combination should be easy

### Overview

1. Define `chat/` `base/` and `team/` modules
2. Define a state solution, `store/` which uses pure functions to update state
    - `createStore(reducer, initialState)`
    - `store.getState()`
    - `store.dispatch(action())`
    - `withStore()`
    - `withGuards()`
3. Define function `withGuards([guardFn1, guardFn2])` that takes an arbitrary number of guard functions and waits for all to resolve.
4. Define a helper `withState(store, withGuards(guards))(wrappedApiFunc)` which composes and `apiFunc` to be able to use `store` without changing its function signature. If `guards` is provided, then all guards must resolve before calling `apiFunc`.
5. Lots of Flow type updates for better typing
6. Update `rollup.config.js` to use Node/NPM modole resolution - this means that it will use `index.js` as the default export of a directory.
7. Update JSDOC syntax to use `@namespace` to structure module hierarchy.

### Bot API

The Bot API is mostly the same but included the discussed changed in #18.

```js
const Bot = require('keybase-chat-bot')
const bot = new Bot()
// Base
bot.init({username: '', paperkey: '', verbose: false})
bot.myInfo()
bot.deinit()

// Chat
bot.chat.list( ... )
bot.chat.read( ... )
bot.chat.send( ... )
bot.chat.delete( ... )
bot.chat.watchChannelForNewMessages( ... )
bot.chat.watchAllChannelsForNewMessages( ... )
```

However when Bots are initialized they create a functional `store` and pass that to each API function using `withStore`.
I kept object instantiation to maintain the same API for now. The alternative would have been `const newBot = require('keybase-chat-bot')` and that seemed a bit off.


### Tests

With API functions being able to have closure over `store`, we will be able to write tests that do the following:

1. Test a bot in any state by defining the state as an object
2. When running tests we can test any API function without having to call `init()` first because we can use `withStore()` and pass a testing store in.
3. Start with any testing store, run `N` API commands, and check to make sure the store is correct
4. Create two stores, two bots that communicate together,  and run `N` API commands on bot 1 and `N` API commands on bot 2 and check to make sure that the store is correct.
